hypothesis-python/src/hypothesis/strategies/_internal/types.py

```diff
@@ -88,10 +88,44 @@ def type_sorting_key(t):
 
 
 def try_issubclass(thing, superclass):
-    thing = getattr(thing, "__origin__", None) or thing
-    superclass = getattr(superclass, "__origin__", None) or superclass
     try:
-        return issubclass(thing, superclass)
+        # Else we're looking at two distinct classes - which might be generics.
+        # That brings in some complications:
+        if issubclass(
+            getattr(thing, "__origin__", None) or thing,
+            getattr(superclass, "__origin__", None) or superclass,
+        ):
+            superclass_args = getattr(superclass, "__args__", None)
+            if sys.version_info[:2] == (3, 6) or not superclass_args:
```

Comment:
I think you can check for the generic args in Python 3.6 as well -- perhaps you make here an assumption from the caller that I am not aware of in which case you might want to document it here?

I tried this on Python 3.6 and it worked for me:
https://github.com/mristin/icontract-hypothesis/commit/ecbfc3a2316cd0d7fa9bc6b904f8ad6d904ebe0c#diff-8d2ae45dc380bf8f4a8627f051f5cd55138de3c4129391b86ba33e846b8276c6R682

If you consider `try_is_subclass` in isolation, I think that you can make it work for Python 3.6 as well.

Reply:
__args__ exists in Python 3.6, but __orig_bases_ doesn't so we can't do the remaining logic.

Reply:
Oh, I see, there is no way to explore the class hierarchy upwards on Python 3.6. Sorry, I missed that -- perhaps you want to explain that in a comment as well?